### ISSUE TITLE
Rule proposal: slot-syntax rule

### ISSUE THREAD

**Issue by ota-meshi** [vuejs/eslint-plugin-vue#800]
**Please describe what the rule should do:**

This rule reports deprecated or removed (in Vue.js 2.6.0) slot syntax.

This rule has the following options.

```json
{
  "vue/slot-syntax": ["error", {
    "v-slot": true,
    "slot": true,
    "slot-scope": true,
    "scope": false
  }]
}
```

- `v-slot` ... If `false`, disallow `v-slot`. default `true`
- `slot` ... If `false`, disallow `slot`. default `true`
- `slot-scope` ... If `false`, disallow `slot-scope`. default `true`
- `scope` ... If `false`, disallow `scope`. default `false`

I think we need to discuss the default options.
The listed defaults are designed not to warn v2.5 - 2.6 users.

**What category should the rule belong to?**

- [X] Enforces code style
- [ ] Warns about a potential error
- [X] Suggests an alternate way of doing something
- [ ] Other (please specify:)

**Provide 2-3 code examples that this rule should warn about:**

```vue
<template>
  <ListComponent>
    <!-- ✓ GOOD -->
    <template v-slot:name="props">
      {{ props.title }}
    </template>
    <template slot="name" slot-scope="props">
      {{ props.title }}
    </template>
  </ListComponent>
  <ListComponent>
    <!-- ✗ BAD -->
    <template slot="name" scope="props">
      {{ props.title }}
    </template>
  </ListComponent>
</template>
```

- option `{"v-slot": true, "slot": false, "slot-scope": false, "scope": false}`

> You are using Vue.js v2.6.0+ and you want to use only `v-slot`

```vue
<template>
  <ListComponent>
    <!-- ✓ GOOD -->
    <template v-slot:name="props">
      {{ props.title }}
    </template>
  </ListComponent>
  <ListComponent>
    <!-- ✗ BAD -->
    <template slot="name" slot-scope="props">
      {{ props.title }}
    </template>
    <template slot="name" scope="props">
      {{ props.title }}
    </template>
  </ListComponent>
</template>
```

- option `{"v-slot": false, "slot": true, "slot-scope": true, "scope": false}`

> You are using Vue.js v2.5.x

```vue
<template>
  <ListComponent>
    <!-- ✓ GOOD -->
    <template slot="name" slot-scope="props">
      {{ props.title }}
    </template>
  </ListComponent>
  <ListComponent>
    <!-- ✗ BAD -->
    <template v-slot:name="props">
      {{ props.title }}
    </template>
    <template slot="name" scope="props">
      {{ props.title }}
    </template>
  </ListComponent>
</template>
```

**Additional context**

[v-slot](https://vuejs.org/v2/api/#v-slot)
[slot](https://vuejs.org/v2/api/#slot-deprecated)
[slot-scope](https://vuejs.org/v2/api/#slot-scope-deprecated)
[scope](https://vuejs.org/v2/api/#scope-removed)

---

I started implementing on [this branch](https://github.com/vuejs/eslint-plugin-vue/tree/add-rule/slot-syntax).
However, we need to make the parser supports to `v-slot`.

**Comment by mysticatea** [comment]
I like this idea.

I guess that this rule's name should start with `no-` because the purpose is to disallow legacy syntax. For example, `vue/no-deprecated-api` or something like.

For too new syntax, I'm imaging another rule such as `vue/no-unsupported-features`, similar to [node/no-unsupported-features/es-syntax](https://github.com/mysticatea/eslint-plugin-node/blob/b234dcc6a653f8b74fbf1f2773b71bcb06b7b255/docs/rules/no-unsupported-features/es-syntax.md).

**Comment by ota-meshi** [comment]
@mysticatea Thank you for comment!
If we add the following rules, will I add a separate js file for each syntax check like`node/no-unsupported-features/es-syntax`?
- `vue/no-deprecated-api`
- `vue/no-unsupported-features`

**Comment by ota-meshi** [comment]
@mysticatea 
I will add the deprecated check rules with the following names.
- `vue/no-deprecated-slot-attribute`
- `vue/no-deprecated-slot-scope-attribute`
- `vue/no-deprecated-scope-attribute`

In addition, I will add the rule to check the new syntax with the name `vue/no-unsupported-features` and let the user set the version. What do you think?

**Comment by mysticatea** [comment]
Sounds good to me!
